subs/pantry/src/Pantry/Storage.hs

```diff
@@ -436,30 +449,121 @@ loadHackageTarballInfo name version = do
+    let filePathId :: FilePathId = entityKey filepath
+    hpackTreeEntry <-
+        selectFirst [TreeEntryTree ==. tid, TreeEntryPath ==. filePathId] []
+    hpackEntity <-
+        case hpackTreeEntry of
+            Nothing ->
+                error $
+                "generateHPack: No package.yaml file found in TreeEntry for TreeId:  " ++
+                (show tid)
+            Just record -> return record
+    hpackBS <- loadBlobById (treeEntryBlob $ entityVal hpackEntity)
+    (pkgName, cabalBS) <- lift $ hpackToCabal hpackBS
+    bid <- storeCabalFile cabalBS pkgName
+    let cabalFile = P.cabalFileName pkgName
+    fid <- insertBy FilePath {filePathPath = cabalFile}
+    let hpackRecord =
+            HPack
+                { hPackTree = tid
+                , hPackVersion = vid
+                , hPackCabal = bid
+                , hPackPath = getKey fid
+                }
+    val <- insertBy hpackRecord
+    return $ getKey val
+
+
+hpackVersionId ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => ReaderT SqlBackend (RIO env) VersionId
+hpackVersionId = do
+    hpackSoftwareVersion <- lift $ hpackVersion
+    hid <- insertBy $ Version {versionVersion = P.VersionP hpackSoftwareVersion}
+    return $ getKey hid
+
+storeHPack ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => TreeId
+    -> ReaderT SqlBackend (RIO env) (Key HPack)
+storeHPack tid = do
+    vid <- hpackVersionId
+    hpackRecord <- getBy (UniqueHPack tid vid)
+    case hpackRecord of
+      Nothing -> generateHPack tid vid
+      Just record -> return $ entityKey record
+
 storeTree
-  :: (HasPantryConfig env, HasLogFunc env)
+  :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
   => P.PackageIdentifier
   -> P.Tree
   -> P.TreeEntry
   -- ^ cabal file
+  -> P.BuildFileType
   -> ReaderT SqlBackend (RIO env) (TreeId, P.TreeKey)
-storeTree (P.PackageIdentifier name version) tree@(P.TreeMap m) (P.TreeEntry (P.BlobKey cabal _) cabalType) = do
+storeTree (P.PackageIdentifier name version) tree@(P.TreeMap m) (P.TreeEntry (P.BlobKey btypeSha _) fileType) buildFileType = do
   (bid, blobKey) <- storeBlob $ P.renderTree tree
-  mcabalid <- loadBlobBySHA cabal
-  cabalid <-
-    case mcabalid of
-      Just cabalid -> pure cabalid
-      Nothing -> error $ "storeTree: cabal BlobKey not found: " ++ show (tree, cabal)
+  buildTypeid <- loadBlobBySHA btypeSha
+  buildid <-
+    case buildTypeid of
+      Just buildId -> pure buildId
+      Nothing -> error $ "storeTree: " ++ (show buildFileType) ++ " BlobKey not found: " ++ show (tree, btypeSha)
   nameid <- getPackageNameId name
   versionid <- getVersionId version
+  let cabalid' = case buildFileType of
+                   P.HPackFile -> Nothing
+                   P.CabalFile -> Just buildid
   etid <- insertBy Tree
     { treeKey = bid
-    , treeCabal = cabalid
-    , treeCabalType = cabalType
+    , treeCabal = cabalid'
```

Comment:
If you implement the `data BuildFile` idea I mentioned elsewhere, then these two fields will collapse into one.